### PR TITLE
[Merged by Bors] - fix issues with too many point lights

### LINKED ISSUE #3915: too many point lights behave strangely
## Bevy version

main

## Operating system & version

Windows 10

## What you did

Create a scene with more than 256 point lights.

## What you expected to happen

Rational degradation of the lights in the scene.

## What actually happened

Odd behaviour as some clusters get more light than expected, and some get less / none.

## Additional information

Example code to see the issue (most noticeable after ~10 seconds when the CLUSTER_COUNT_MASK limit is exceeded for the middle cluster):

```rust
use bevy::{prelude::*, diagnostic::{FrameTimeDiagnosticsPlugin, LogDiagnosticsPlugin}, window::{WindowMode, PresentMode}};

fn main() {
    App::new()
        .insert_resource(WindowDescriptor{
            width: 1280.0,
            height: 720.0,
            mode: WindowMode::Windowed,
            present_mode: PresentMode::Immediate,
            ..Default::default()
        })
        .add_plugins(DefaultPlugins)
        .add_plugin(FrameTimeDiagnosticsPlugin)
        .add_plugin(LogDiagnosticsPlugin::default())
        .add_startup_system(setup)
        .add_system(move_camera)
        .run()
}

fn setup(
    mut commands: Commands,
    mut meshes: ResMut<Assets<Mesh>>,
    mut materials: ResMut<Assets<StandardMaterial>>,
) {
    // some quad
    commands.spawn_bundle(PbrBundle{
        mesh: meshes.add(shape::Quad{size: Vec2::splat(100.0), flip: false}.into()),
        material: materials.add(Color::WHITE.into()),
        transform: Transform::from_translation(Vec3::new(0.0, 0.0, -5.0)),
        ..Default::default()
    });

    // some bright offscreen lights
    for x in 0..25 {
        for y in 0..10 {
            commands.spawn_bundle(PointLightBundle {
                point_light: PointLight {
                    color: Color::RED,
                    intensity: 10.0,
                    range: f32::sqrt(10.0 * 10.0 / (4.0 * std::f32::consts::PI)),
                    ..Default::default()
                },
                transform: Transform::from_translation(Vec3::new(20.0 + 10.0 * x as f32, y as f32 * 10.0 - 50.0, -4.9)),
                ..Default::default()
            });
        }
    }

    // grid of lights
    for x in 1..=18 {
        for y in 1..=18 {
            commands.spawn_bundle(PointLightBundle {
                point_light: PointLight {
                    color: Color::BLUE,
                    intensity: x as f32 * y as f32 * 0.01,
                    range: f32::sqrt(x as f32 * y as f32 * 0.01 * 10.0 / (4.0 * std::f32::consts::PI)),
                    ..Default::default()
                },
                transform: Transform::from_translation(Vec3::new(x as f32 * 0.2 - 2.0, y as f32 * 0.2 - 2.0, -4.9)),
                ..Default::default()
            });
        }
    }

    //camera
    commands.spawn_bundle(PerspectiveCameraBundle::new_3d());

    //action
}

fn move_camera(
    mut q: Query<&mut Transform, With<Camera>>,
    time: Res<Time>,
) {
    for mut t in q.iter_mut() {
        t.translation.z += time.delta_seconds();
    }
}

```
lights_issue_1
lights_issue_2

### PR DESCRIPTION
**Pull request by robtfm**
# Objective

fix #3915 

## Solution

the issues are caused by
- lights are assigned to clusters before being filtered down to MAX_POINT_LIGHTS, leading to cluster counts potentially being too high
- after fixing the above, packing the count into 8 bits still causes overflow with exactly 256 lights affecting a cluster

to fix:

```assign_lights_to_clusters```
- limit extracted lights to MAX_POINT_LIGHTS, selecting based on shadow-caster & intensity (if required)
- warn if MAX_POINT_LIGHT count is exceeded

```prepare_lights```
- limit the lights assigned to a cluster to CLUSTER_COUNT_MASK (which is 1 less than MAX_POINT_LIGHTS) to avoid overflowing into the offset bits

notes:
- a better solution to the overflow may be to use more than 8 bits for cluster_count (the comment states only 14 of the remaining 24 bits are used for the offset). this would touch more of the code base but i'm happy to try if it has some benefit.
- intensity is only one way to select lights. it may be worth allowing user configuration of the light filtering, but i can't see a clean way to do that

### CODE REVIEW

File: crates/bevy_pbr/src/light.rs [hunk 1]
```diff
@@ -506,7 +554,7 @@ pub fn assign_lights_to_clusters(
             vec![VisiblePointLights::from_light_count(light_count); cluster_count];
         let mut visible_lights_set = HashSet::with_capacity(light_count);
 
-        for (light_entity, light_transform, light) in lights.iter() {
+        for &(light_entity, light_transform, light) in lights.iter() {
```
Comment (superdump) [inline crates/bevy_pbr/src/light.rs:557]: Just curious why is this change needed?

**Review by superdump — COMMENTED** [review]
Just the one question, otherwise LGTM. When the question is answered, I'll mark it as ready for final review.

File: crates/bevy_pbr/src/light.rs (hunk 1 above)
Comment (robtfm) [reply crates/bevy_pbr/src/light.rs:557]: This is because we're iterating the Vec instead of the Query so there's another indirection. Could use into_iter() instead since we don't use the vec again.

File: crates/bevy_pbr/src/light.rs [hunk 2]
```diff
@@ -480,14 +481,61 @@ fn cluster_to_index(cluster_dimensions: UVec3, cluster: UVec3) -> usize {
...
+                frusta
+                    .iter()
+                    .any(|frustum| frustum.intersects_sphere(&light_sphere))
+            })
+            // take one extra light to check if we should emit the warning
+            .take(MAX_POINT_LIGHTS + 1)
+            .collect();
+
+        if lights.len() > MAX_POINT_LIGHTS {
+            warn!("MAX_POINT_LIGHTS exceeded");
```
Comment (superdump) [inline crates/bevy_pbr/src/light.rs:533]: It would be good to note what the MAX_POINT_LIGHTS limit is to make this more actionable.

File: crates/bevy_pbr/src/light.rs (hunk 1 above)
Comment (superdump) [reply crates/bevy_pbr/src/light.rs:557]: I guess if they aren't used again then into iter is maybe better?

**Review by superdump — CHANGES_REQUESTED** [review]
A couple of small changes then LGTM.

File: crates/bevy_pbr/src/light.rs (hunk 1 above)
Comment (robtfm) [reply crates/bevy_pbr/src/light.rs:557]: on trying to change i realised this is actually used in the per-view loop, so we can't use into_iter after all.

File: crates/bevy_pbr/src/light.rs (hunk 2 above)
Comment (robtfm) [reply crates/bevy_pbr/src/light.rs]: done

File: crates/bevy_pbr/src/light.rs (hunk 1 above)
Comment (superdump) [reply crates/bevy_pbr/src/light.rs:557]: Ok, no problem then.

**Review by superdump — APPROVED** [review]

File: crates/bevy_pbr/src/light.rs [hunk 3]
```diff
@@ -480,14 +481,61 @@ fn cluster_to_index(cluster_dimensions: UVec3, cluster: UVec3) -> usize {
...
+
 // NOTE: Run this before update_point_light_frusta!
 pub fn assign_lights_to_clusters(
     mut commands: Commands,
     mut global_lights: ResMut<VisiblePointLights>,
     mut views: Query<(Entity, &GlobalTransform, &Camera, &Frustum, &mut Clusters)>,
     lights: Query<(Entity, &GlobalTransform, &PointLight)>,
 ) {
-    let light_count = lights.iter().count();
+    let mut lights = lights.iter().collect::<Vec<_>>();
```
Comment (mockersf) [inline crates/bevy_pbr/src/light.rs:504]: I think only the translation is used from the transform. would it make sense to only keep that before collecting?

File: crates/bevy_pbr/src/light.rs [hunk 4]
```diff
@@ -480,14 +481,61 @@ fn cluster_to_index(cluster_dimensions: UVec3, cluster: UVec3) -> usize {
...
+                frusta
+                    .iter()
+                    .any(|frustum| frustum.intersects_sphere(&light_sphere))
+            })
+            // take one extra light to check if we should emit the warning
+            .take(MAX_POINT_LIGHTS + 1)
+            .collect();
+
+        if lights.len() > MAX_POINT_LIGHTS {
+            warn!("MAX_POINT_LIGHTS ({}) exceeded", MAX_POINT_LIGHTS);
```
Comment (mockersf) [inline crates/bevy_pbr/src/light.rs:533]: not a fan of logs that will potentially happens on every frame

File: crates/bevy_pbr/src/light.rs (hunk 3 above)
Comment (robtfm) [reply crates/bevy_pbr/src/light.rs:504]: we use everything - note that this lights variable shadows the query (otherwise it'd be better to remove it from the query).

File: crates/bevy_pbr/src/light.rs (hunk 4 above)
Comment (robtfm) [reply crates/bevy_pbr/src/light.rs:533]: i agree, but it should definitely be emitted once. 

is it reasonable to add a Local<bool> to check if the warning has been emitted already?

Comment (superdump) [reply crates/bevy_pbr/src/light.rs:533]: The problem is that if there are other logs then it may get lost. I think we need a general pattern for this. Perhaps something like a Local<> that holds the last time we logged, and then just log once every 10s or something?

File: crates/bevy_pbr/src/light.rs (hunk 3 above)
Comment (mockersf) [reply crates/bevy_pbr/src/light.rs:504]: I mean something like 
```suggestion
    let mut lights = lights.iter().map(|(e, gt, pl)| (e, gt.translation, pl)).collect::<Vec<_>>();
```
scale and rotation from the lights are not used. I'm not sure of the perf impact, but it would change the footprint from `Vec3 + Vec3 + Quat` to just `Vec3`

File: crates/bevy_pbr/src/light.rs (hunk 4 above)
Comment (mockersf) [reply crates/bevy_pbr/src/light.rs:533]: no, I would keep it that way for now rather than add an additional parameter... there are a few other places with similar issues, we don't have the good solution for that yet

Comment (mockersf) [reply crates/bevy_pbr/src/light.rs:533]: ideally the logger should be able to detect duplicate logs and only log them once in a while 🤔

File: crates/bevy_pbr/src/light.rs (hunk 3 above)
Comment (robtfm) [reply crates/bevy_pbr/src/light.rs]: sorry, you're quite right. i've done that now.

**Review by mockersf — APPROVED** [review]

File: crates/bevy_pbr/src/light.rs [hunk 5]
```diff
@@ -480,14 +481,64 @@ fn cluster_to_index(cluster_dimensions: UVec3, cluster: UVec3) -> usize {
...
+
 // NOTE: Run this before update_point_light_frusta!
 pub fn assign_lights_to_clusters(
     mut commands: Commands,
     mut global_lights: ResMut<VisiblePointLights>,
     mut views: Query<(Entity, &GlobalTransform, &Camera, &Frustum, &mut Clusters)>,
     lights: Query<(Entity, &GlobalTransform, &PointLight)>,
 ) {
-    let light_count = lights.iter().count();
+    let mut lights = lights
```
Comment (cart) [inline crates/bevy_pbr/src/light.rs:504]: Given that these are potentially large per-frame allocations, I think we should be reusing the same vecs via  `Local` system params.

File: crates/bevy_pbr/src/light.rs (hunk 4 above)
Comment (cart) [reply crates/bevy_pbr/src/light.rs:536]: For something that should be "fixed" i'm cool with warning every frame (until we can suppress more intelligently). But in this case, complex scenes could easily go over the max point light limit. This isn't "wrong" and I dont think we should be aggressively pushing people to hide lights that go over the limit.

Manually suppressing duplicate warnings is important to do here imo.

Comment (superdump) [reply crates/bevy_pbr/src/light.rs:536]: @cart what do you think of my suggestion above as an approach to 'rate limiting' the log?

Comment (robtfm) [reply crates/bevy_pbr/src/light.rs:555]: limited to warning once for now

File: crates/bevy_pbr/src/light.rs (hunk 5 above)
Comment (robtfm) [reply crates/bevy_pbr/src/light.rs]: done - i also extracted only the necessary parts of the point light data to reduce the size a bit more.

File: crates/bevy_pbr/src/light.rs (hunk 4 above)
Comment (cart) [reply crates/bevy_pbr/src/light.rs:553]: @superdump  yeah i like that, although I'm cool with logging once for now. "rate limiting" is an option, but we could also consider detecting fluctuations in light counts (ex: if you go under the limit and then back over, we print the warning again).

File: crates/bevy_pbr/src/light.rs [hunk 6]
```diff
@@ -478,14 +479,85 @@ fn ndc_position_to_cluster(
...
+        lights.sort_by(|light_1, light_2| {
+            point_light_order(
+                (&light_1.entity, &light_1.shadows_enabled),
+                (&light_2.entity, &light_2.shadows_enabled),
+            )
+        });
+
+        // check each light against each view's frustum, keep only those that affect at least one of our views
+        let frusta: Vec<_> = views.iter().map(|(_, _, _, frustum, _)| *frustum).collect();
+        *lights = (*lights)
```
Comment (cart) [inline crates/bevy_pbr/src/light.rs:535]: We could use `retain` here to avoid the extra allocation (combine the filter predicate with a captured counter to limit to MAX_POINT_LIGHTS + 1 and remove the remaining lights)

Comment (robtfm) [reply crates/bevy_pbr/src/light.rs]: good idea, thanks. the checks are failing now but i think it's not my fault?

Comment (cart) [reply crates/bevy_pbr/src/light.rs]: Yup its an github linux-vm CI problem. Bleh :)
